File: pine-java/src/test/java/page/liam/pine/ItemColumnTest.java
package page.liam.pine;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for OperatorInput.itemColumn: element i must be identical to
 * item(i, field) across both storage modes, materialized mode, defaults
 * substitution and absent fields. Mirrors pine-go's item_column_test.go.
 */
class ItemColumnTest {

    private static List<Map<String, Object>> sampleItems() {
        List<Map<String, Object>> items = new ArrayList<>();
        Map<String, Object> r0 = new LinkedHashMap<>();
        r0.put("a", 1.0);
        r0.put("b", "x");
        items.add(r0);
        Map<String, Object> r1 = new LinkedHashMap<>();
        r1.put("a", 2.0); // b missing
        items.add(r1);
        Map<String, Object> r2 = new LinkedHashMap<>();
        r2.put("a", null);
        r2.put("b", "z");
        items.add(r2);
        return items;
    }

    private static InputFieldSpec spec(List<String> nullableItem,
                                       List<InputFieldSpec.DefaultedField> defaultedItem) {
        return new InputFieldSpec(
                new ArrayList<>(), new ArrayList<>(), new ArrayList<>(),
                new ArrayList<>(), defaultedItem, nullableItem);
    }

    private static void forBothModes(FrameConsumer consumer) throws Exception {
        for (String mode : new String[]{"row", "column"}) {
            consumer.accept(mode, Frame.create(mode, new HashMap<>(), sampleItems()));
        }
    }

    interface FrameConsumer {
        void accept(String mode, Frame frame) throws Exception;
    }

    @Test
    void itemColumnMatchesItem() throws Exception {
        forBothModes((mode, frame) -> {
            List<String> nullable = new ArrayList<>();
            nullable.add("a");
            OperatorInput in = frame.buildInput("op", spec(nullable, new ArrayList<>()));
            for (String field : new String[]{"a", "b", "absent"}) {
                Object[] col = in.itemColumn(field);
                assertEquals(in.itemCount(), col.length, mode + ": field " + field);
                for (int i = 0; i < col.length; i++) {
                    assertEquals(in.item(i, field), col[i],
                            mode + ": field " + field + " item " + i);
                }
            }
        });
    }

    @Test
    void itemColumnAppliesDefaults() throws Exception {
        List<Map<String, Object>> items = new ArrayList<>();
        Map<String, Object> r0 = new LinkedHashMap<>();
        r0.put("score", 1.0);
        items.add(r0);
        Map<String, Object> r1 = new LinkedHashMap<>();
        r1.put("score", null);
        items.add(r1);
        items.add(new LinkedHashMap<>()); // missing entirely

        for (String mode : new String[]{"row", "column"}) {
            Frame frame = Frame.create(mode, new HashMap<>(), items);
            List<InputFieldSpec.DefaultedField> defaulted = new ArrayList<>();
            defaulted.add(new InputFieldSpec.DefaultedField("score", -1.0));
            OperatorInput in = frame.buildInput("op", spec(new ArrayList<>(), defaulted));
            Object[] col = in.itemColumn("score");
            Object[] want = new Object[]{1.0, -1.0, -1.0};
            for (int i = 0; i < want.length; i++) {
                assertEquals(want[i], col[i], mode + ": item " + i);
                assertEquals(want[i], in.item(i, "score"), mode + ": item() " + i);
            }
        }
    }

    @Test
    void itemColumnMaterializedMode() {
        List<Map<String, Object>> items = new ArrayList<>();
        Map<String, Object> r0 = new LinkedHashMap<>();
        r0.put("a", 1.0);
        items.add(r0);
        Map<String, Object> r1 = new LinkedHashMap<>();
        r1.put("a", 2.0);
        items.add(r1);
        OperatorInput in = new OperatorInput(null, items);
        Object[] col = in.itemColumn("a");
        assertEquals(2, col.length);
        assertEquals(1.0, col[0]);
        assertEquals(2.0, col[1]);
    }

    @Test
    void itemColumnViewWindow() {
        List<Map<String, Object>> items = new ArrayList<>();
        for (int i = 0; i < 10; i++) {
            Map<String, Object> row = new LinkedHashMap<>();
            row.put("v", (double) i);
            items.add(row);
        }
        for (String mode : new String[]{"row", "column"}) {
            Frame frame = Frame.create(mode, new HashMap<>(), items);
            Object[] col = frame.itemColumnView("v", 3, 4);
            assertNotNull(col, mode);
            assertEquals(4, col.length, mode);
            for (int i = 0; i < 4; i++) {
                assertEquals((double) (3 + i), col[i], mode + ": slot " + i);
            }
            assertNull(frame.itemColumnView("v", 8, 4), mode + ": out-of-range window");
            assertNull(frame.itemColumnView("v", -1, 2), mode + ": negative offset");
        }
    }

    @Test
    void itemColumnAbsentField() throws Exception {
        forBothModes((mode, frame) -> {
            OperatorInput in = frame.buildInput("op", spec(new ArrayList<>(), new ArrayList<>()));
            Object[] col = in.itemColumn("nope");
            assertEquals(in.itemCount(), col.length, mode);
            for (Object v : col) {
                assertNull(v, mode);
            }
        });
    }
}

File: pine-go/internal/dataframe/item_column_test.go
package dataframe

import (
	"fmt"
	"testing"

	"github.com/Liam0205/pineapple/pine-go/internal/config"
	"github.com/Liam0205/pineapple/pine-go/internal/types"
)

// itemColumnModes runs a subtest against both storage modes.
func itemColumnModes(t *testing.T, fn func(t *testing.T, mode StorageMode)) {
	t.Helper()
	for _, tm := range testModes {
		t.Run(tm.name, func(t *testing.T) { fn(t, tm.mode) })
	}
}

func TestItemColumnMatchesItem(t *testing.T) {
	itemColumnModes(t, func(t *testing.T, mode StorageMode) {
		items := []map[string]any{
			{"a": float64(1), "b": "x"},
			{"a": float64(2)}, // b missing
			{"a": nil, "b": "z"},
		}
		f := NewFrame(mode, map[string]any{"u": "c"}, items)
		in, err := f.BuildInput("op", &config.InputFieldSpec{
			NullableItem: []string{"a"},
		})
		if err != nil {
			t.Fatal(err)
		}
		for _, field := range []string{"a", "b", "absent"} {
			col := in.ItemColumn(field)
			if len(col) != in.ItemCount() {
				t.Fatalf("field %q: len=%d want %d", field, len(col), in.ItemCount())
			}
			for i := range col {
				want := in.Item(i, field)
				if fmt.Sprint(col[i]) != fmt.Sprint(want) {
					t.Errorf("field %q item %d: got %v want %v", field, i, col[i], want)
				}
			}
		}
	})
}

func TestItemColumnAppliesDefaults(t *testing.T) {
	itemColumnModes(t, func(t *testing.T, mode StorageMode) {
		items := []map[string]any{
			{"score": float64(1)},
			{"score": nil},
			{}, // missing entirely
		}
		f := NewFrame(mode, nil, items)
		in, err := f.BuildInput("op", &config.InputFieldSpec{
			DefaultedItem: []config.DefaultedField{{Name: "score", Default: float64(-1)}},
		})
		if err != nil {
			t.Fatal(err)
		}
		col := in.ItemColumn("score")
		want := []any{float64(1), float64(-1), float64(-1)}
		for i := range want {
			if col[i] != want[i] {
				t.Errorf("item %d: got %v want %v", i, col[i], want[i])
			}
			if got := in.Item(i, "score"); got != want[i] {
				t.Errorf("Item(%d): got %v want %v", i, got, want[i])
			}
		}
	})
}

func TestItemColumnMaterializedMode(t *testing.T) {
	items := []map[string]any{
		{"a": float64(1)},
		{"a": float64(2)},
	}
	in := types.NewOperatorInput(nil, items)
	col := in.ItemColumn("a")
	if len(col) != 2 || col[0] != float64(1) || col[1] != float64(2) {
		t.Fatalf("materialized ItemColumn = %v", col)
	}
}

func TestItemColumnViewWindow(t *testing.T) {
	items := make([]map[string]any, 10)
	for i := range items {
		items[i] = map[string]any{"v": float64(i)}
	}
	itemColumnModes(t, func(t *testing.T, mode StorageMode) {
		f := NewFrame(mode, nil, items)
		cr, ok := f.(types.ColumnReader)
		if !ok {
			t.Fatalf("%T does not implement ColumnReader", f)
		}
		col, ok := cr.ItemColumnView("v", 3, 4)
		if !ok {
			t.Fatal("ItemColumnView not ok")
		}
		if len(col) != 4 {
			t.Fatalf("len=%d want 4", len(col))
		}
		for i := 0; i < 4; i++ {
			if col[i] != float64(3+i) {
				t.Errorf("slot %d: got %v want %v", i, col[i], float64(3+i))
			}
		}
		// Out-of-range windows must be rejected.
		if _, ok := cr.ItemColumnView("v", 8, 4); ok {
			t.Error("out-of-range window unexpectedly ok")
		}
		if _, ok := cr.ItemColumnView("v", -1, 2); ok {
			t.Error("negative offset unexpectedly ok")
		}
	})
}

func TestItemColumnAbsentField(t *testing.T) {
	itemColumnModes(t, func(t *testing.T, mode StorageMode) {
		f := NewFrame(mode, nil, []map[string]any{{"a": 1.0}, {"a": 2.0}})
		in, err := f.BuildInput("op", &config.InputFieldSpec{})
		if err != nil {
			t.Fatal(err)
		}
		col := in.ItemColumn("nope")
		if len(col) != 2 {
			t.Fatalf("len=%d want 2", len(col))
		}
		for i, v := range col {
			if v != nil {
				t.Errorf("slot %d: got %v want nil", i, v)
			}
		}
	})
}
